pkg/glusterfs/controllerserver.go

```diff
@@ -0,0 +1,377 @@
+	}
+	glog.V(2).Infof("Deleting volume with ID: %v", req.VolumeId)
+
+	err := cs.client.VolumeStop(req.VolumeId)
+
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to stop volume %s", err.Error())
+	}
+
+	err = cs.client.VolumeDelete(req.VolumeId)
+	if err != nil {
+		glog.Errorf("Volume delete failed :%v", err)
+		return nil, status.Errorf(codes.Internal, "error deleting volume: %s", err.Error())
+	}
+	return &csi.DeleteVolumeResponse{}, nil
+}
+
+// ControllerPublishVolume attaches the given volume to the node
+func (cs *ControllerServer) ControllerPublishVolume(ctx context.Context, req *csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+//ControllerUnpublishVolume deattaches the given volume from the node
+func (cs *ControllerServer) ControllerUnpublishVolume(ctx context.Context, req *csi.ControllerUnpublishVolumeRequest) (*csi.ControllerUnpublishVolumeResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "")
+}
+
+// ValidateVolumeCapabilities checks whether the volume capabilities requested
+// are supported.
+func (cs *ControllerServer) ValidateVolumeCapabilities(ctx context.Context, req *csi.ValidateVolumeCapabilitiesRequest) (*csi.ValidateVolumeCapabilitiesResponse, error) {
+
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "volume capabilities request is nil")
+	}
+
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "ValidateVolumeCapabilities() - Volume ID is nil")
+	}
+
+	if req.VolumeCapabilities == nil {
+		return nil, status.Error(codes.InvalidArgument, "ValidateVolumeCapabilities is nil")
+	}
+
+	var vcaps []*csi.VolumeCapability_AccessMode
+	for _, mode := range []csi.VolumeCapability_AccessMode_Mode{
+		csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER,
+	} {
+		vcaps = append(vcaps, &csi.VolumeCapability_AccessMode{Mode: mode})
+	}
+
+	hasSupport := func(mode csi.VolumeCapability_AccessMode_Mode) bool {
+		for _, m := range vcaps {
+			if mode == m.Mode {
+				return true
+			}
+		}
+		return false
+	}
+
+	resp := &csi.ValidateVolumeCapabilitiesResponse{
+		Supported: false,
+	}
+
+	for _, cap := range req.VolumeCapabilities {
+		if hasSupport(cap.AccessMode.Mode) {
+			resp.Supported = true
+		} else {
+			resp.Supported = false
+		}
+	}
+	glog.V(1).Infof("glusterfs CSI driver support capabilities: %v", resp)
+	return resp, nil
+}
+
+// ListVolumes returns a list of all requested volumes
+func (cs *ControllerServer) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (*csi.ListVolumesResponse, error) {
+
+	//Fetch all the volumes in the TSP
+	volumes, err := cs.client.Volumes("")
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	var entries []*csi.ListVolumesResponse_Entry
+	for _, vol := range volumes {
+		v, e := cs.client.VolumeStatus(vol.Name)
```

Comment:
We should implement a `VolumeStatusAll` API in this. It would simplify this so much and make it perform better at the scale of volumes GCS expects. Filed issue gluster/glusterd2#1136 to track this.

Reply:
if we store volume size during the volume creation, we don't need to implement new API in gd2.
I have created an issue to add volume size filed in volume info.  [issue](https://github.com/gluster/glusterd2/issues/1081)

Reply:
@kshlm The `csi.ListVolumesResponse` only needs volume ID and Capacity. All we have to do is add capacity to volume info. We already support listing all volume infos in a single API call